crates/driver/src/domain/competition/solution/mod.rs

```diff
@@ -270,29 +332,93 @@ impl std::fmt::Debug for Solution {
     }
 }
 
+/// Given two solutions returns the factors with
+/// which prices of the second solution would have to be multiplied so that the
+/// given token would have the same price in both solutions.
+/// If the solutions have no prices in common any scaling factor is valid (we
+/// return 1). Returns None if the solutions have more than one price in common
+/// and the scaling factor is not unique.
+fn scaling_factor(first: &Prices, second: &Prices) -> Option<BigRational> {
+    let factors: HashSet<_> = first
+        .keys()
+        .collect::<HashSet<_>>()
+        .intersection(&second.keys().collect::<HashSet<_>>())
+        .map(|&token| {
+            let first_price = first[token];
+            let second_price = second[token];
+            BigRational::new(
+                number::conversions::u256_to_big_int(&second_price),
+                number::conversions::u256_to_big_int(&first_price),
+            )
+        })
+        .collect();
+    match factors.len() {
+        0 => Some(BigRational::one()),
+        1 => factors.into_iter().next(),
+        _ => None,
+    }
+}
+
 /// Carries information how the score should be calculated.
 #[derive(Debug, Clone)]
 pub enum SolverScore {
     Solver(eth::U256),
     RiskAdjusted(f64),
 }
+
+impl SolverScore {
+    fn merge(&self, other: &SolverScore) -> Option<SolverScore> {
+        match (self, other) {
+            (SolverScore::Solver(a), SolverScore::Solver(b)) => Some(SolverScore::Solver(a + b)),
+            (SolverScore::RiskAdjusted(a), SolverScore::RiskAdjusted(b)) => {
+                Some(SolverScore::RiskAdjusted(a * b))
+            }
+            _ => None,
+        }
+    }
+}
+
 /// A unique solution ID. This ID is generated by the solver and only needs to
 /// be unique within a single round of competition. This ID is only important in
 /// the communication between the driver and the solver, and it is not used by
 /// the protocol.
-#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq)]
-pub struct Id(pub u64);
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+pub enum Id {
+    Single(u64),
+    Merged(Vec<u64>),
+}
 
 impl From<u64> for Id {
     fn from(value: u64) -> Self {
-        Self(value)
+        Id::Single(value)
     }
 }
 
-impl From<Id> for u64 {
-    fn from(value: Id) -> Self {
-        value.0
+impl Id {
+    /// Returns the number of solutions that has gone into merging this
+    /// solution.
+    pub fn count_merges(&self) -> usize {
+        self.ids().len()
     }
+
+    pub fn ids(&self) -> Vec<u64> {
+        match self {
+            Id::Single(id) => vec![*id],
+            Id::Merged(ids) => ids.clone(),
+        }
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum MergeError {
```

Comment:
In some of the previous PRs, I started moving these error types into inner `error` module, so we could refer to it with `error::Merge` instead of `MergeError`.